### PR TITLE
feat: create_download_link tool with one-time HTTP download URLs

### LINKED ISSUE #69: feat: create_download_link tool with one-time artifact HTTP endpoint
## Problem

MCP has a hub-and-spoke architecture — there is no server-to-server channel. When an image is generated by this server and needs to be consumed by another MCP server (e.g., `markdown-vault-mcp` saving an attachment, a Slack MCP posting an image), the binary data must round-trip through the client's context window. This is impractical for full-resolution images (1-10 MB).

## Background

Discussed in S2S session (2026-03-20). Three approaches were considered:
1. **Shared Docker volume** — only works same-host, tight coupling between specific servers
2. **In-process composition** — `mcp.mount()` shares memory but tightly couples servers
3. **One-time HTTP download URLs** — server exposes images as short-lived HTTP endpoints with OTP tokens

Approach 3 was selected. FastMCP's HTTP transport already runs on uvicorn/Starlette, so adding a route is trivial. The consuming MCP server (e.g., `markdown-vault-mcp`) would implement a `save_artifact_from_url(url, path)` tool that fetches the URL server-to-server.

**Key design decisions:**
- Token is the auth — no bearer/OIDC on the artifact endpoint (it must bypass MCP auth)
- In-memory token store — does not need to survive server restart
- Configurable TTL (default 5 minutes) — unclaimed tokens expire
- Hidden on `--transport stdio` — no HTTP server available, tool is meaningless
- Accepts a resource URI (e.g., `image://abc123/view?format=webp&width=512`) — transforms encoded in the URI
- `IMAGE_GENERATION_MCP_BASE_URL` is reused to construct the full download URL

## Design Documentation

No design documents cover artifact download endpoints. Implementation decisions in this issue become the de facto specification. An ADR should be considered if the pattern proves reusable.

## Acceptance Criteria

- [ ] 1. **New `create_download_link` tool** in `_server_tools.py` with parameters:
  - `uri` (str, required): Resource URI (e.g., `image://abc123/view?format=webp&width=512`)
  - `ttl_seconds` (int, optional, default 300): Time before the token expires
- [ ] 2. Tool returns `TextContent` with JSON: `{"download_url": "https://host/artifacts/{token}", "expires_in_seconds": 300, "uri": "image://..."}`
- [ ] 3. Tool validates that the resource URI references a registered image (raises error for unknown image IDs)
- [ ] 4. **Artifact HTTP endpoint** at `GET /artifacts/{token}` mounted on the Starlette app alongside the MCP endpoint
- [ ] 5. Endpoint serves the resolved image bytes with correct `Content-Type` header, then **deletes the token** (one-time use)
- [ ] 6. Expired tokens return HTTP 404 (not 401/403 — no information leakage about token validity)
- [ ] 7. Unknown/already-used tokens return HTTP 404
- [ ] 8. **In-memory token store**: `dict[str, TokenRecord]` where `TokenRecord` holds `uri`, `created_at`, `ttl_seconds`. Cleanup of expired tokens on each request (lazy expiry) or periodic sweep.
- [ ] 9. **Token generation**: `uuid4()` — 122 bits of entropy, sufficient for short-lived OTP
- [ ] 10. **Hidden on stdio transport**: Tool is not registered when `--transport stdio`. Implementation: check transport mode during `register_tools()` or use a tag that is disabled for stdio.
- [ ] 11. **Base URL required**: Tool raises a clear error if `IMAGE_GENERATION_MCP_BASE_URL` is not configured
- [ ] 12. **Artifact endpoint bypasses MCP auth**: The route must not be behind bearer/OIDC middleware. The OTP token is the sole authentication.
- [ ] 13. `IMAGE_GENERATION_MCP_BASE_URL` documentation updated — it is now a general server config, not OIDC-specific. Update `docs/configuration.md`, `README.md` config table, and `docs/guides/authentication.md` to reflect broader scope.

## Test Updates

- Add unit tests for token store: create, retrieve, expire, double-use returns None
- Add unit tests for `create_download_link` tool: valid URI, invalid image ID, missing BASE_URL
- Add integration test for artifact endpoint: serve bytes, verify one-time use, verify expired token 404
- Add test that tool is not registered on stdio transport
- Patch coverage must not regress below 80%

## Out of Scope

- Consuming side (`save_artifact_from_url`) — that belongs in `markdown-vault-mcp` or other consuming servers
- Persistent token store (database/Redis) — in-memory is sufficient for homelab
- IP-based access restriction — OTP + TTL is sufficient auth
- HTTPS/TLS on the artifact endpoint — handled by reverse proxy (Traefik)

## Dependencies

- Blocked by: #67 (ResourceLink contract and resource URI pattern must be established first)

## Definition of Done

- [ ] All acceptance criteria checked off
- [ ] Tests pass, coverage not regressed
- [ ] architect-reviewer conformance table in PR body
- [ ] `docs/configuration.md` updated with BASE_URL broader scope
- [ ] `docs/tools.md` updated with `create_download_link` tool

### PR DESCRIPTION
**Pull request by pvliesdonk**
## Summary

- New `create_download_link` tool that issues one-time HTTP download URLs for server-to-server image transfer
- New `artifacts.py` module — `ArtifactStore` backed by `OrderedDict`, with configurable TTL and lazy expiry cleanup
- Token is invalidated on first download (one-time use)
- Tool is transport-conditional: hidden on `stdio` transport, only active on HTTP/SSE
- Artifact endpoint mounted via `mcp.custom_route()` and bypasses MCP auth

## Changes

- `artifacts.py` (NEW, 240 lines) — `ArtifactStore`, `TokenRecord`, `make_artifact_handler`
- `_server_tools.py` — `create_download_link` tool with transport guard
- `_server_deps.py` — `ArtifactStore` injected into lifespan
- `mcp_server.py` — `custom_route()` wiring for artifact endpoint
- `cli.py`, `config.py` — `ARTIFACT_TTL_SECONDS` env var
- `docs/tools.md`, `docs/design/provider-system.md`, `docs/configuration.md` — updated
- `tests/test_artifacts.py` (449 lines, 25 tests)

## Not Included / Future PRs

- Persistent artifact store (disk-backed) — not needed for current use case
- Multi-use download links — not in scope

## Design Conformance

| # | Requirement | Source | Status | Evidence |
|---|---|---|---|---|
| 1 | One-time download URL with token invalidation | docs/tools.md create_download_link | CONFORMANT | artifacts.py consume_token() |
| 2 | Configurable TTL with lazy expiry cleanup | docs/tools.md | CONFORMANT | artifacts.py _cleanup_expired() |
| 3 | Transport-conditional registration (hidden on stdio) | docs/tools.md, design doc | CONFORMANT | _server_tools.py transport check |
| 4 | Requires BASE_URL env var | docs/tools.md, docs/configuration.md | CONFORMANT | _server_tools.py:363-370 |
| 5 | Artifact endpoint bypasses MCP auth | docs/tools.md | CONFORMANT | mcp_server.py custom_route |
| 6 | Docs updated (tools, design, configuration) | CLAUDE.md doc rules | CONFORMANT | all doc files updated |

Reviewed by @architect-reviewer — all requirements CONFORMANT.

## Test Plan

- `test_artifacts.py` — 25 tests: token creation, one-time consumption, TTL expiry, lazy cleanup, concurrent access, missing token 404

## Risk / Rollback

- Additive: no existing tools modified except adding new route
- `BASE_URL` not set → tool registers but raises at call time with a clear error
- Tokens held in-memory; server restart clears all pending download links

Fixes #69

### CODE REVIEW

File: src/image_generation_mcp/artifacts.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        store = get_artifact_store()
+        record = store.consume_token(token)
+
+        if record is None:
+            logger.debug("Artifact token not found or expired: %s", token)
+            return Response(content="Not Found", status_code=404)
+
+        # Parse the resource URI to extract image_id and transform params
+        parsed = urlparse(record.uri)
+        image_id = parsed.hostname or parsed.netloc
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/image_generation_mcp/artifacts.py]: Fixed in f32ce6f. Replaced `parsed.hostname` with `parsed.netloc` to preserve case.

File: tests/test_artifacts.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+        import image_generation_mcp._server_deps as deps_mod
+        from image_generation_mcp.artifacts import (
+            ArtifactStore,
+            make_artifact_handler,
+            set_artifact_store,
+        )
+
+        store = ArtifactStore()
+        set_artifact_store(store)
+        original_service = deps_mod._service_store
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply tests/test_artifacts.py]: Intentionally kept as-is. The test helper creates a fresh service per test class via the `service` fixture, so cross-test pollution is bounded. Adding monkeypatch-based teardown would make the helper harder to reuse across test classes.

File: src/image_generation_mcp/_server_deps.py [hunk 3]
```diff
@@ -75,9 +101,19 @@ async def _service_lifespan(
         # Discover capabilities for all registered providers
         await service.discover_all_capabilities()
 
+        # Populate module-level store for artifact handler access
+        _service_store = service
+
+        # Initialise artifact store
+        from image_generation_mcp.artifacts import ArtifactStore, set_artifact_store
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/image_generation_mcp/_server_deps.py]: Intentionally kept as-is. These imports are inside functions to avoid circular imports — the modules reference each other's types. Moving them to top-level would require restructuring the dependency graph.

File: src/image_generation_mcp/mcp_server.py [hunk 4]
```diff
@@ -260,10 +265,20 @@ def create_server() -> FastMCP:
...
     )
 
-    register_tools(mcp)
+    register_tools(mcp, transport=transport)
     register_resources(mcp)
     register_prompts(mcp)
 
+    # Mount artifact download endpoint for HTTP transports
+    if transport != "stdio":
+        from image_generation_mcp.artifacts import make_artifact_handler
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/image_generation_mcp/mcp_server.py]: Same as above — kept inside conditional block to avoid importing when the feature is unused (stdio transport).
